src/librustc/traits/error_reporting.rs

```diff
@@ -378,12 +379,16 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
             flags.push(("_Self".to_string(), Some(self.tcx.type_of(def.did).to_string())));
         }
 
-        for param in generics.types.iter() {
+        for param in generics.params.iter() {
+            let value = match param.kind {
+                GenericParamDefKind::Type(_) => {
+                    let ty = trait_ref.substs.type_for_def(&param);
+                    ty.to_string()
```

Comment:
I wonder if this can be made to work with `trait_ref.substs[param.index].to_string()`.